lib/Promises/Deferred.pm

```diff
@@ -4,186 +4,207 @@ package Promises::Deferred;
-
-    $self->{'result'} = [@_];
-    $self->{'status'} = RESOLVED;
-    $self->_notify;
-    $self;
-}
-
-sub reject {
-    my $self = shift;
-    die "Cannot reject. Already  " . $self->status
-        unless $self->is_in_progress;
-
-    $self->{'result'} = [@_];
-    $self->{'status'} = REJECTED;
-    $self->_notify;
-    $self;
+        # Explicit undef. Why? Because this is where we're going to spend almost all of our time:
+        # deallocating Perl objects. So make it explicit as profilers will point people here.
+        undef $cb;
+    }
 }
 
-sub then {
-    my $self = shift;
-    my ( $callback, $error ) = $self->_callable_or_undef(@_);
-
-    my $d = ( ref $self )->new;
-    push @{ $self->{'resolved'} } => $self->_wrap( $d, $callback, 'resolve' );
-    push @{ $self->{'rejected'} } => $self->_wrap( $d, $error,    'reject' );
-
-    $self->_notify unless $self->is_in_progress;
-    $d->promise;
+sub new {
+    return bless \{cb=>[], state=>0}, __PACKAGE__;
 }
 
-sub chain { 
-    my $self = shift;
-    $self = $self->then($_) for @_;
+sub deferred (;&) {
+    my $self= new();
+    if (my $code= shift) {
+        $self->resolve;
+        return $self->then(sub{
+            $code->($self);
+        });
+    }
     return $self;
 }
 
-sub catch {
-    my $self = shift;
-    $self->then( undef, @_ );
-}
+sub _invoke_cbs {
+    my ($cbs)= @_;
+    return unless @$cbs;
 
-sub done {
-    my $self = shift;
-    my ( $callback, $error ) = $self->_callable_or_undef(@_);
-    push @{ $self->{'resolved'} } => $callback if defined $callback;
-    push @{ $self->{'rejected'} } => $error    if defined $error;
+    my $should_schedule= !@pending_callbacks;
+    push @pending_callbacks, @$cbs;
 
-    $self->_notify unless $self->is_in_progress;
-    ();
+    if ($should_schedule) {
+        $notify_sub->();
+    }
 }
 
-sub finally {
-    my $self = shift;
-    my ($callback) = $self->_callable_or_undef(@_);
-
-    my $d = ( ref $self )->new;
+sub _invoke_cb {
+    my $cb= shift;
+    my $self= shift @$cb;
 
-    if (defined $callback) {
-        my ( @result, $method );
-        my $finish_d = sub { $d->$method(@result); () };
-
-        my $f = sub {
-            ( $method, @result ) = @_;
-            local $@;
-            my ($p) = eval { $callback->(@result) };
-            if ( $p && blessed $p && $p->can('then') ) {
-                return $p->then( $finish_d, $finish_d );
+    if (my $invoke_callback= $cb->[$$self->{state}]) {
+        eval {
+            my @result= $invoke_callback->(@{$$self->{result}});
+            if (my $next= $cb->[0]) {
+                if (@result == 1 && is_blessed_ref($result[0]) && $result[0]->can('then')) {
+                    if ($result[0]->isa("Promises::Promise")) {
+                        _replace_promise($next, ${$result[0]});
+                    } elsif ($result[0]->isa("Promises::Deferred")) {
+                        _replace_promise($next, $result[0]);
+                    } else {
```

Comment:
So looks like the ref-of-ref instance is causing issues here, these conditionals feel/smell wrong.